### PR TITLE
Use protocompile as the compiler

### PR DESCRIPTION
**Pull request by jhump**
This replaces the prior compiler (`github.com/jhump/protoreflect/desc/protoparse`) with a new one in `github.com/bufbuild/protocompile`.

- Weave build options through more of the API since there are now more flags that external callers are interested in than just "exclude source info".
- Update `buf build` and `buf generate` so they use descriptors that match those of protoc byte-for-byte.
- Update `buf lint` to generate *extra* comments, since the default source info from protocompile matches protoc. The level of comments generated by protoc is too low for linting, which wants extra comments (as captured by prior compiler, protoparse) for checking for nolint annotation comments.

Fixes TCN-294

### CODE REVIEW

File: private/bufpkg/bufimage/bufimagebuild/builder.go
Comment (jhump) [inline private/bufpkg/bufimage/bufimagebuild/builder.go:91]: Much of the red lines in this file are because `protocompile` natively handles concurrency. So a lot of logic in here that merged chunks of built results is no longer needed.

File: private/bufpkg/bufimage/bufimagebuild/builder.go
Comment (jhump) [inline private/bufpkg/bufimage/bufimagebuild/builder.go:191]: `protocompile` _is_ concurrent. However, this explicit locking is unnecessary as it handles thread-safe access to the given reporter internally.
https://pkg.go.dev/github.com/bufbuild/protocompile@v0.0.0-20221004142453-757ef5688631/reporter#Reporter

File: CHANGELOG.md
Comment (bufdev) [inline CHANGELOG.md:15]: Make sure we update before merging

File: go.mod
Comment (bufdev) [inline go.mod:14]: We should file an issue to rework the test in `bufimageutil_test.go` to not use `protoprint`, which I think is possible given what I see. A good CLI onboarding issue probably.

File: make/buf/all.mk
Comment (bufdev) [inline make/buf/all.mk:1]: Update to `https://github.com/bufbuild/protocompile/commits/main DATE_OF_LAST_COMMIT checked DATE_WE_LAST_CHECKED_FOR_A_NEW_COMMIT`

File: private/buf/cmd/buf/command/build/build.go
Comment (bufdev) [inline private/buf/cmd/buf/command/build/build.go:151]: I'm not sure we want to lock ourselves to the SourceCodeInfos that protoc produces - there's no exact definition as to what SourceCodeInfos a FileDescriptorSet will produce, and us producing extra is arguably better than what protoc does. I would not change the SourceCodeInfos dependent on the state of `--as-file-descriptor-set`, instead I'd remove the `WithExtraSourceCodeInfo` option altogether and make having extra source code info the default.

File: private/buf/cmd/buf/command/build/build.go
Comment (bufdev) [inline private/buf/cmd/buf/command/build/build.go:155]: Effectively same comment - I don't think we need to lock ourselves long-term into emitting the exact same bytes as protoc. We fulfill the contract that protoc has by producing functionally-equivalent FileDescriptorSets, our definition of what these are is just as good.

File: private/buf/cmd/buf/command/generate/generate.go
Comment (bufdev) [inline private/buf/cmd/buf/command/generate/generate.go:317]: Regardless of the answers to the above, why do we need to do this for `buf generate`?

File: private/bufpkg/bufcheck/bufbreaking/bufbreaking_test.go
Comment (bufdev) [inline private/bufpkg/bufcheck/bufbreaking/bufbreaking_test.go:176]: I'm concerned that any of these changed. If anything, I'd expect that part of our test for if this protoparse-to-protocompile conversion worked would be that this file would have no diff. Can you explain here a bit?

File: private/bufpkg/bufcheck/bufbreaking/testdata/breaking_field_same_json_name/1.proto
Comment (bufdev) [reply private/bufpkg/bufcheck/bufbreaking/testdata/breaking_field_same_json_name/1.proto:10]: I think the is the answer to the above question, let me know.

File: private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go
Comment (bufdev) [inline private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go:62]: See comment above, I don't think we should have this option.

File: private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go
Comment (bufdev) [inline private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go:71]: Is this different than before? All parallelism in bufbuild/buf should be controlled by `pkg/thread`, so whatever `pkg/thread` has set for parallelism should be what we use within `bufimagebuild`.

File: private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go
Comment (bufdev) [inline private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go:84]: See comment above, I don't think we should have this option, we should choose what the correct answer is (canonical or not-canonical) and always do that. Happy to discuss offline.

File: private/bufpkg/bufimage/bufimagebuild/builder.go
Comment (bufdev) [inline private/bufpkg/bufimage/bufimagebuild/builder.go:343]: Why where these removed? I'm likely just not looking carefully enough, but `bufimagebuild` needs to produce a full `Image`, which includes these extra things that `FileDescriptorSets` do not have.

File: private/buf/cmd/buf/command/build/build.go
Comment (jhump) [reply private/buf/cmd/buf/command/build/build.go:151]: > there's no exact definition as to what SourceCodeInfos a FileDescriptorSet will produce

It is documented in `descriptor.proto`. And `protoparse` wasn't really conforming to it (intentionally -- I originally added the extra comments before I implemented the AST, because I wanted a way to get access to as much of the original source as possible).

I'm fine removing it. But it seemed valuable for maximizing trust that a descriptor set produced by buf matches that of protoc.

File: private/buf/cmd/buf/command/build/build.go
Comment (jhump) [reply private/buf/cmd/buf/command/build/build.go:155]: This was basically the fix for https://github.com/jhump/protoreflect/issues/214

If you don't think we should do that, it's easy to remove. But I thought there was value in it.

File: CHANGELOG.md
Comment (jhump) [reply CHANGELOG.md:15]: @bufdev, you could "request changes" on the PR, and that will block merging. (Or some other reviewer could do that.)

File: private/bufpkg/bufcheck/bufbreaking/bufbreaking_test.go
Comment (jhump) [reply private/bufpkg/bufcheck/bufbreaking/bufbreaking_test.go:176]: I had to change JSON names for the files to actually compile. I added dumb number suffixes, which made some of the spans grow by one.

File: private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go
Comment (jhump) [reply private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go:71]: I'll rework it. I originally added this before discovering the `thread` package and seeing how it was used before. That is in fact how we set the _default_ parallelism. The only caller of this is a test (that sets it to 1 for deterministic output). I can instead have the test set the parallelism via the `thread` package (and defer to restore to prior setting).

File: private/buf/cmd/buf/command/build/build.go
Comment (bufdev) [reply private/buf/cmd/buf/command/build/build.go:151]: Let's discuss tomorrow or in person this week

File: private/buf/cmd/buf/command/build/build.go
Comment (bufdev) [reply private/buf/cmd/buf/command/build/build.go:155]: Let's discuss tomorrow/in person this week

File: private/bufpkg/bufimage/bufimagebuild/builder.go
Comment (pkwarren) [inline private/bufpkg/bufimage/bufimagebuild/builder.go:334]: The old protoreflect would just return non-public, non-weak dependencies with `GetDependencies()`. Do we need to filter those out here in this loop?

Comment (jhump) [reply private/bufpkg/bufimage/bufimagebuild/builder.go:334]: This isn't right. The old protoreflect returns all dependencies. The other accessors for public and weak return a subset of this.
https://pkg.go.dev/github.com/jhump/protoreflect@v1.13.0/desc#FileDescriptor.GetDependencies

File: .github/dependabot.yml
Comment (jhump) [reply .github/dependabot.yml]: fixed

File: private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go
Comment (jhump) [reply private/bufpkg/bufimage/bufimagebuild/bufimagebuild.go]: Removed this option.

File: CHANGELOG.md
Comment (jhump) [reply CHANGELOG.md]: Updated with more details and actual stats. Related: https://github.com/bufbuild/protocompile/pull/64

File: private/bufpkg/bufcheck/bufbreaking/testdata/breaking_field_same_json_name/1.proto
Comment (jhump) [reply private/bufpkg/bufcheck/bufbreaking/testdata/breaking_field_same_json_name/1.proto:10]: Yep.

**Comment by jhump** [comment]
Per convo with @bufdev, I've updated this PR to (1) always use "standard" source code info (same level of comments as protoc) and (2) not bother trying to emit "canonical" bytes for descriptor sets. We can easily roll either these back in if we change our minds. For the first item (amount of comments), I had to update a test since the linter can no longer access a comment on the individual request or response type definition in a method.

Since we don't need ways to customize these aspects of output, I removed the relevant build options and reverted a lot of the method signature changes (since those changes are no longer necessary/useful).

cc @pkwarren

**Review by bufdev — APPROVED** [review]

File: CHANGELOG.md
Comment (jhump) [reply CHANGELOG.md]: Moved it up.

**Review by pkwarren — APPROVED** [review]

**Pull request merged**
